### ISSUE TITLE
Automating training of models for best results

### ISSUE THREAD

**Issue by 1N50MN14** [gafr/chainer-fast-neuralstyle-models#8]
Hi guys, first of all thanks for the awesome work!

I'm new to all of this and I'm not sure exactly sure which parameters would yield the best results for each style. Also went through the comments, based on what I understood, also from @6o6o comment on the starry night model there's a lot of trial and error involved.

So I have this idea (I'm relatively good at distributing work across clusters, spawning machines on demands); if we could pick X number popular styles (from Prisma or similar) and someone with more experience could indicate possible ranges for each training parameter (say for example we get 125 possible combinations for a single style) I could dynamically launch 125 * X instances and get them to do the training, report back to us (email, upload whatever) and shut them down when finished. 

Yeah one time potentially high cost (I'll take it on me) but then we would've filled up this repo. Just a thought, what do you think?

**Comment by Daddaaff** [comment]
I'd love to see that! @6o6o should give some hints about what parameters to play with. 

You should start with a specific style and find out how to remove those dots and borders.

**Comment by 6o6o** [comment]
Hi, thanks for the offer.

As you have noted, there's no general way to make everything look great. I've trained dozens of styles, but only a few turned out good. You need to explore many options and be prepared for some fail rate. Going for styles that already work well in many of today's apps will increase chances of success.

Generally, the style needs to have hard, sharp, well-defined features with lots of edges. Details of all sorts of different shapes and angles should be present to better match the variety of objects in content images. Color diversity is also important. At least 3-4 well-distinguished colors should be included. I'd say skin tones are a must, unless it is desired to repaint a skin in a picture. And even then, every time I think I found the the right components, some style comes up, which fails spectacularly and makes me rethink my approach :)

As for parameters, there's not much space here, at least for what I know. You can try increasing total variation `lambda_tv` to 10e-3 or 10e-2. It tends to produce deeper colors and smooth out excessive details. Playing with `lambda_feat` and `lambda_style` may also be useful, but note, it's only the relation between the two that matters, so you should fix one and tweak the other.

If you have further questions, let me know. I can provide you with small list of styles I always wanted to train, but never had the time.

Thanks.

**Comment by 1N50MN14** [comment]
@6o6o Thank you for getting back to me! ;)

Yes I think it's more efficient for me to get a list of styles from you that you believe would produce good results, i.e. a list of styles that we know work in many of today's apps as you mentioned. It would be even a greater help for me if it was possible for you to somehow "guess" the relationships between `lambda_feat`and `lambda_style` and note the CLI params for each style, this might save us some time, and save me some $ as every second counts on so many machines ;)

This way I can run all of them in one batch, send back the results to you for evaluation, then I can make another run for the ones you think can be improved. Do you think you have some spare time (not much, just evaluate the output and adjust the CLI params) to help me out with this?

**Comment by 6o6o** [comment]
I think the default settings are the required way to go first. Then, if you generally like the style, but would like to tweak some settings (washed out colors, grainy details, etc.), you may adjust parameters or alter the input style itself. I tend to make checkpoints every 20k images to see the early stages of the outcome, since it's not very feasible to wait 20 hours per epoch just to get a glimpse of it.

What hardware are you planning to run this on? The amazon g2.2xlarge instance features a Kepler Tesla K10, which is a bit outdated by today's standards and takes about 2 days per epoch.

**Comment by 1N50MN14** [comment]
Yeah I saw that, I'll basically be looking up a good provider from a list which I found here: http://www.nvidia.com/object/gpu-cloud-computing-services.html

As long as they bill by the hour so I'm not charged for a whole month...

**Comment by 1N50MN14** [comment]
https://pod.penguincomputing.com/pricing <- Tesla K40 (12 GB mem), think this would do?

**Comment by 6o6o** [comment]
Do they charge only for working hours? Or the time the server stands idle is counted as well?

Give me some time to populate the list and I'll send it to you around the same time tomorrow.

**Comment by 1N50MN14** [comment]
I'm not sure but I think they would also charge for idle time and stop charging only when the machine is shut down... I can first try on a single machine see how that goes, then run a whole batch on multiple machines.

No worries, appreciate the help (feel free to send the styles by email if it's easier for you)

**Comment by 1N50MN14** [comment]
By the way, how intensive is it to apply a pre-trained module on images if I may ask? I'm going to attempt to apply that on a video (say 15 seconds, 450 frames at 30fps) in under 5 seconds, is that possible with a CPU machine or would it require/make use of good GPU machines as well?

**Comment by 6o6o** [comment]
I think the K10 on g2.2xlarge is just fine if it offers a better price / performance ratio. Time is not crucial here, we can wait, unless you're in a hurry )

The 720p frame would take around 20 sec to process on a quad-core CPU and < 1 sec on GPU. Of course, if you downscale the image to 640x360 first, it would take about 5 seconds.

**Comment by 1N50MN14** [comment]
Not so much in a hurry, I just got the feeling waiting 24 hours is going to frustrate the heck out of me :) Going to do the math, also found http://www.cirrascale.com/cloud/plans.aspx by the way, looks a bit expensive does pre-training harness multiple GPU's?

My frame size is 450x800 so I guess that's comparable to <1 sec on GPU per frame; so for 450 frames that's 450 seconds? it's the g2.2xlarge GPU you're referring to correct?

**Comment by 6o6o** [comment]
No, the training code does not utilize multiple GPUs.

450x800 pic should require about 6 sec on quad-core CPU to generate.

**Comment by 6o6o** [comment]
Here are the first four images. The first two, _Art Nouveau_ and _Blue Dream_ were selected from Artisto app and should give acceptable results for general audience.

The second pair, _Modern Art_ and _Spiral_ was my personal choice, will likely result in a heavy transformation and would be suitable for people with more sophisticated taste )

Use code in my repo for training: https://github.com/6o6o/chainer-fast-neuralstyle
After setting everything up, download MS COCO dataset and run the command
```
python train.py -s <style_image_path> -d <train2014_dataset_path> -g 0
```
[styles.zip](https://github.com/gafr/chainer-fast-neuralstyle-models/files/454093/styles.zip)

**Comment by 1N50MN14** [comment]
Cool thank you, I'm on it - give me couple of days to figure out the right machine / vendor / price to use and I will run those four; after this batch we can try more styles (including all of the ones you like :) )

**Comment by 6o6o** [comment]
No problem man. There's no hurry. Thank you for helping in the first place.

**Comment by rupeshs** [comment]
These styles can be used with Neuralstyler 2.0
http://neuralstyler.com/styles.html

**Comment by dovanchan** [comment]
How about the project now?Where can I find the model trained
